### ISSUE TITLE
Future of sections

### ISSUE THREAD

**Issue by SkySkimmer** [coq/coq#6254]
I think the discussion in #883 should be continued in an issue.

I'll quote @herbelin as an opening post:

> Hi, I remain unclear about how to make progresses on section variables (and, incidentally, on this PR). I did not re-read carefully all comments but the general idea seems to be well-represented by this comment from @silene: « the intuitive behavior is to clear these variables. It is just that, until all the bugs related to clearing section variables are fixed, we should refrain from automatically clearing section variables by default. »
> 
> So, how to step further and to eventually address the clearing section variables bugs, one way or another? Shall we think at a different model, e.g. based on contexts? With which transition strategy?
> 
> Shall we first move sections out of the kernel with on-the-fly generalization of the constants? But with which semantics for clearing a section variable? Forbidden? Allowed, but cutting all dependencies of the cleared variables in global objects?
> 
> More generally, is the dependency of notations, implicit arguments, hints, etc. in section variables a problem, or, on the opposite, we would like to eventually associate notations, or implicit arguments, etc. also to goal variables (think e.g. at a local polymorphic function which we just posed in the context and to which we would like to set implicit arguments and define an infix notation).
> 
> Depending on what we want, different models can be considered. For instance, we could invent a systematic way to refer to goal variables which were cleared (since after all, they are still there somewhere in the spine of lambdas of the current proof being constructed). Or, to control the effect of a clear, we could imagine that we maintain a table telling for each goal variable in which other objects it depends in.
> 
> So, shortly, it seems to me that it is first a lot about what user model we'd like to eventually propose, and that once the model is clear the technical constraints will resolve by themselves.
> 
> Any comments?

**Comment by SkySkimmer** [comment]
Personally I don't understand what contexts are meant to be.

I think we want something in the environment to be able to decide if `A` is a section variable or an evar variable. This could be separating the name context into a section and an evar context, but I also considered the following idea:
Suppose we change constr so it goes `type (...) kind = ... | Var of variable | ...` where `variable = Id.t * tag` and `tag = int option`. Tag `None` means a section variable, tag `Some i` means an evar variable with the value of `i` being used to fail when a tactic does 
```coq
let x := constr(A (* Some i *)) in
  clear A (*, Some i *); 
  intros A (* , Some fresh j *);
  do_something_with_x (* (A, Some i) not in context (A, Some j) *)
```

In any case once we can tell which section variables are available, if we also tag objects with the ones they use we can detect when they're accessed in an invalid context, and either complain (if accessed explicitly, eg `apply foo` where `foo` is a constant using a cleared variable) or skip ahead silently (eg for hints).

Maybe I'm missing something though, eg how feasible is it to find which variables a hint / tactic / whatever is using, or issues with sections that this doesn't solve / makes harder to solve.

**Comment by ejgallego** [comment]
Let me add a section implementation quiz here. Options have a `write` / `read` functions. Now imagine the following code:
```Coq
Section A.
Set Foo.
End A.
```
When `End A` is processed, what is the `write` trace for `Foo`? What happens if instead of `Section` we use `Module`.

**Comment by herbelin** [comment]
@SkySkimmer: looks interesting.

Trying to understand your example, are you saying that with using a `i` or `j`, we could have a stronger check on situations like:
```coq
Goal True -> False -> False.
intros A; let tac := ltac:(apply A) in clear A; intro A; tac.
```
which currently works but about which we could make stricter and fail?

@ejgallego: Are you asking about what should be the scope of a `Set` wrt `Section` and `Module` boundaries? I don't remember all the abundant developer literature written on that topic but wasn't some conclusion that using the `Local` and `Global` modifiers was providing with a pragmatical way to bypass the respective default behaviors.

**Comment by SkySkimmer** [comment]
That's the idea, although that specific example maybe can't be made stricter with ltac semantics considering
``` coq
Goal False -> False.
  let tac:= ltac:(apply A) in intros A;tac.
Defined.
```
works.
OTOH
``` coq
Ltac tac := apply A.
```
fails when `A` doesn't exist, so plausibly 
```coq
Section foo.
  Variable A:True.
  Ltac tac := apply A.

  Goal False -> False.
    clear A;intros A;tac.
  Defined.
End foo.
```
could be made to fail.

**Comment by ejgallego** [comment]
@herbelin , no, I am asking a way more concrete question (with large practical implications):

What does the current implementation do with regards to writing the option?

**Comment by ppedrot** [comment]
@SkySkimmer I'd advise against basing a design of section semantics on the semantics of free variables in Ltac, which is notably broken.

**Comment by herbelin** [comment]
> What does the current implementation do with regards to writing the option?

Sorry, I may be tired but I'm unsure to understand well your question. If I say that the state is stored at the entry in a section or module block, restored at the end of the section or module, and a couple of declarations replayed depending on some visibility rules, which `Set` may or may not satisfy (*), is it an appropriate answer to your question?

(*) replayed at the end of sections unless `Local`, not replayed at the end of modules, unless `Global`

**Comment by ejgallego** [comment]
> Sorry, I may be tired but I'm unsure to understand well your question. If I say that the state is stored at the entry in a section or module block, restored at the end of the section or module, and a couple of declarations replayed depending on some visibility rules, which Set may or may not satisfy (*), is it an appropriate answer to your question?

Indeed @herbelin you are spot on. At the end of the section, first the option is unset, then set again. However that means we have a critical interaction wrt to state, as each write is done by different core components (and even registration). So in a sense, we have two visibility mechanisms: modules, and sections.

I find that weird, and it means that it is hard to get things out of the summary as long as they are also registered in the libobject visibility system.

I dunno what we could do here, but likely the way to go is to consolidate all visibility handling into a single system.

**Comment by ejgallego** [comment]
Something that indeed I think it would be interesting to have is a more unified treatment of `Axioms` vs `Section Variables` vs `Future Proofs`.

Conceptually, all of them are the same, placeholders for a witness that may be fulfilled at some time in the future. I think the code now makes a few unnecessary distinctions.

**Comment by Zimmi48** [comment]
I think the actors in this discussion might be interested to learn that Lean tried a different approach and is backtracking to the one currently implemented in Coq:

> `parameters` are currently simulated in Lean. For example, when we declare foo in a section with a parameter `A`, `foo` is automatically abstracted and an alias `foo => @foo A` is created. This creates many problems, most of them are documented in the issue tracker. This approach has one advantage: users can use the abstracted (`_root_.foo`) and non-abstracted (`foo`) version simultaneously. Another advantage is that we don't have to type check `foo` more than once in the kernel. That being said, the disadvantages far outweigh the advantages. We plan to go back to the approach used in Lean1 and Coq.

This is extracted from the design document for Lean4 whose link was recently shared by @ejgallego.

**Comment by maximedenes** [comment]
> This is extracted from the design document for Lean4 whose link was recently shared by @ejgallego.

Do you have a pointer? I don't remember seeing it.

**Comment by Zimmi48** [comment]
Note that they say that the approach used in Coq was also used in the first version of Lean. They moved away from it, then came back. I haven't kept the pointer: @ejgallego do you have it?

**Comment by ejgallego** [comment]
The pointer is in issue #6938

**Comment by RalfJung** [comment]
> This creates many problems, most of them are documented in the issue tracker.

Would be interesting to learn which issues they considered bad enough to go back to an approach they previously had moved away from. I looked over their issue tracker but did not find it clear what is so bad.

**Comment by Zimmi48** [comment]
Let's ask Leonardo (@leodemoura).

**Comment by maximedenes** [comment]
I may consider attending, will discuss it with Yves. If I can be there, I'll ask about sections indeed.

**Comment by Kha** [comment]
Hey there, I'm one of the developers of Lean 4. I will be at Lean Together if anyone wants to chat with me there, but to cut a long story short: We have changed our plans and will not have built-in support for section parameters in Lean 4 at all. In the end, their rare usage did not justify complicating so many parts of the system. And when they were used, it always seemed like a poor substitute for a more module- or locale-like system that is actually usable and extensible after closing the section.

A few issues we had with the aliasing model for parameters:
* https://github.com/leanprover/lean/issues/530
* https://github.com/leanprover/lean/issues/1773
* https://github.com/leanprover/lean/issues/1826

These are/were all solvable, but it's death by a thousand paper cuts.

**Comment by maximedenes** [comment]
I'd like to experiment with splitting the context in two:
- The section context
- The proof local context

The first one is a named_context, as usual. It is invariant through a proof. The second one is an ordered map from names to `Declaration of NameDecl.t | SectionVariable of bool`. The boolean says if the variable was hidden from the proof context.

I suggest some semantic changes: `clear` only hides (`assumption` won't pick it up, but `clear x; intros x` will fail if `x` is a section variable). It is forbidden to rename, change the type, etc, of section variables.

WDYT @ppedrot ?

**Comment by ppedrot** [comment]
Looks like an interesting experiment to me. Maybe, instead of a hidden flag, just a set of names that are section hypotheses. Also, you might also split the `Var` term constructor in two, one for section variables and the other for hypotheses. This would solve most of the problems by virtually creating two distinct namespaces.

**Comment by gares** [comment]
Hum, we already have `Const` for global stuff. This is what data in the section is, actually. Eg if you put `Inductive A` in a section, `A` is not a `Var` (but just `Ind`). I don't know why (ind if) section variables are not declared as (temporary) axioms (if they were, `Var` would already be used for the proof context only).

**Comment by ppedrot** [comment]
This doesn't play well with universe polymorphism though, one can have polymorphic constants but no polymorphic variables. And this would require more support from the kernel to handle cooking. Which is fine BTW, if we have to rehaul sections, we'll definitely have to rewrite that part.

**Comment by gares** [comment]
You are saying that using (temporary) Axioms for Section Variables would make it possible to have universe polymorphic section variables, that is not the case today. (just rephrasing, since I'm not sure I get where your remark is aiming)

**Comment by gares** [comment]
> And this would require more support from the kernel to handle cooking

only if you say you enable section variables to be polymorphic. This is not implied by the fact that you represent them with Axiom (which can, but don't need to be polymorphic).

**Comment by ppedrot** [comment]
> only if you enable section variables to be polymorphic

No, not only. You cannot abstract over an arbitrary set of constants, you need to be careful about dependencies, opacities and the like. So you need to implement this kind of check during cooking, which is not necessary today because variables enforce this kind of invariant statically.

**Comment by RalfJung** [comment]
The main thing I am worried about here (based on issues @robbertkrebbers ran into in Iris) is tactics not being able to treat all variables uniformly.  Things like `destruct x` not removing `x` when it is a section variable are a big problem when writing robust automation.

**Comment by maximedenes** [comment]
@RalfJung to make sure we are on the same page, what do you mean by "removing"? I.e. what impact should it have?

**Comment by RalfJung** [comment]
By "remove" I mean "remove from the context", as in make it no longer be in the context... not sure which other word to use here? When `x: bool`, after `destruct x` there is no more `x` if and only if `x` was a non-section variable.

That's what I remember, anyway. When I just tried, it wouldn't show section variables in the context at all... wtf.

**Comment by ppedrot** [comment]
> tactics not being able to treat all variables uniformly

You can also add the naming scheme of bound variable introduced by refinement. Depending on whether a variable is a section variable or not, the naming is not the same.

**Comment by maximedenes** [comment]
> remove from the context

Just to make explicit your actual need, what would be the problem if the variable stayed there? I'm aware that you'd like `assumption` and automation tactics not to pick it up. What else?

**Comment by robbertkrebbers** [comment]
> Just to make explicit your actual need, what would be the problem if the variable stayed there? I'm aware that you'd like `assumption` and automation tactics not to pick it up. What else?

That's indeed an important point. Also, it would be more consistent if the following were to work:

```coq
Section foo.
  Context (x : nat).

  Goal True.
  destruct x as [|x].
  (* x is already used. *)
```

**Comment by gares** [comment]
FTR, If `x` was a (temporary) axiom, then it would work (without the need of clearing x, since the `x` being destructed would not be part of the proof context, but just "printed as such" for backward compatibility, and "masked" by the new `x` being introduced in the proof context).

**Comment by RalfJung** [comment]
> what would be the problem if the variable stayed there? I'm aware that you'd like assumption and automation tactics not to pick it up. What else?

Another problem is something like `repeat match goal with H: something |- _ => destruct ... end`. This relies on `destruct` cleaning up or else it will diverge because it destructs the same thing over and over.

**Comment by maximedenes** [comment]
> Another problem is something like `repeat match goal with H: something |- _ => destruct ... end`. This relies on `destruct` cleaning up or else it will diverge because it destructs the same thing over and over.

Well, in fact, is it so clear that section variables should be part of the proof context? Global definitions (or axioms) are not, and section variables are locally global objects (global until they are abstracted over).

**Comment by RalfJung** [comment]
That would be a much more drastic departure from the status quo -- for example, then TC search should also not pick up section variables.

If done consistently, that would also avoid these issues we mentioned. If it would overall work better, I do not know.

**Comment by SkySkimmer** [comment]
IIRC instance section variables get picked up because of some special 
handling of Auto Instance for section variables, not because they're in 
the context

Gaëtan Gilbert

**Comment by gares** [comment]
> Well, in fact, is it so clear that section variables should be part of the proof context?

When I say that section stuff should be global I don't mean that it should not be displayed in the proof context or not kept into account by TC resolution, or not available to ltac/ltac2/elpi/mtac/. I'm just saying that pretending it is part of the proof context is not in line with how the system is actually implemented, so the illusion breaks often and badly (eg stream of bugs related to clear or rename). Of course the question "can a section variable be cleared/renamed" becomes moot, since they would not be variables anymore (and you can't clear or rename, say, `nat`).

The section context could be displayed "Iris style", and be seen as an immutable prefix of the proof context (printed in order to give a meaning to the variables used in the rest of the goal)
```
A : Type
------------
x : A
=======
?evar[x] + 1 = 2
```
Finally, I think that one could imagine to have `match goal with` but also `match section with` in order to be able to express proof search machinery as powerful as before. clear, as a mean to use linearly things, would work for the proof context only.

**Comment by gares** [comment]
Just in case, in non ltac1 proof languages, the context better first class, eg a list, so one can just filter it in order to remove things, no need to call the `clear` tactic.
